pkg/sriov/sriov.go

```diff
@@ -78,31 +82,29 @@ func (s *sriovManager) SetupVF(conf *sriovtypes.NetConf, podifName string, netns
 		return fmt.Errorf("error setting temp IF name %s for %s", tempName, linkName)
 	}
 
-	// Save the original effective MAC address before overriding it
-	conf.OrigVfState.EffectiveMAC = linkObj.Attrs().HardwareAddr.String()
-	// 3. Set MAC address
-	if conf.MAC != "" {
-		err = utils.SetVFEffectiveMAC(s.nLink, tempName, conf.MAC)
-		if err != nil {
-			return fmt.Errorf("failed to set netlink MAC address to %s: %v", conf.MAC, err)
-		}
-	}
-
-	// 4. Change netns
+	// 3. Change netns
 	if err := s.nLink.LinkSetNsFd(linkObj, int(netns.Fd())); err != nil {
 		return fmt.Errorf("failed to move IF %s to netns: %q", tempName, err)
 	}
 
 	if err := netns.Do(func(_ ns.NetNS) error {
-		// 5. Set Pod IF name
+		// 4. Set Pod IF name
 		if err := s.nLink.LinkSetName(linkObj, podifName); err != nil {
 			return fmt.Errorf("error setting container interface name %s for %s", linkName, tempName)
 		}
 
-		// 6. Enable IPv4 ARP notify and IPv6 Network Discovery notify
+		// 5. Enable IPv4 ARP notify and IPv6 Network Discovery notify
 		// Error is ignored here because enabling this feature is only a performance enhancement.
 		_ = s.utils.EnableArpAndNdiscNotify(podifName)
 
+		// 6. Set MAC address
```

Comment:
Oh I did not catch this in U/S sriov-cni. But setting the MAC address after we send out GARPs may be an issue. Since ARPs require that the MAC address is setup properly.

You are moving `SetVFEffectiveMAC` down to solve a race condition, is that correct? So moving it before `EnableArpAndNdiscNotify` shouldn't be any different (although solving race conditions this way really is a toss of the coin).

Reply:
> Oh I did not catch this in U/S sriov-cni. But setting the MAC address after we send out GARPs may be an issue. Since ARPs require that the MAC address is setup properly.

Does the GARP go out before setting up the interface? That happens on step 7

> You are moving SetVFEffectiveMAC down to solve a race condition, is that correct? So moving it before EnableArpAndNdiscNotify shouldn't be any different (although solving race conditions this way really is a toss of the coin).

I agree, that's the best we could do from the CNI side while the investigation on the kernel side is still ongoing.
As the root cause is not yet clear, moving that statement can make a difference. At least, we should run an extensive test suite as we did with these changes.

Reply:
Sorry, I forgot that it is also done after IPAM. Then this is lgtm.

Reply:
It actually doesn't happen at step 7 but after IPAM is setup.